Test package detection in a systematic way (#18175)
LLNL/spack: lib/spack/docs/packaging_guide.rst

.. _package-lifecycle:

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Add detection tests to packages
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

To ensure that a software is detected correctly for multiple configurations
and on different systems users can write a ``detection_test.yaml`` file and
put it in the same directory as the corresponding ``package.py`` file.
This YAML file contains enough information for Spack to mock an environment
and try to check if the detection logic yields the results that are expected.

> As a general rule, attributes at the top-level of ``detection_test.yaml``

Reviewer: I'm concerned about having a domain-specific language vs. adding functions to the package itself -  why not have the user write a function in `package.py`?

```
def test_inspections(self):
  bin1 = 'bin/clang-3.9'
  script1 = """\
echo "clang version 3.9.1-19ubuntu1 (tags/RELEASE_391/rc2)"
echo "Target: x86_64-pc-linux-gnu"
echo "Thread model: posix"
echo "InstalledDir: /usr/bin"
"""
  bin2 = ...
  result2 = ...

  specs = ['llvm@3.9.1 +clang~lld~lldb']

  return [(bin1, result1), (bin2, result2)], specs
```

Author: I don't see this as a domain specific language, more like an additional YAML file that packagers can provide to specify tests for package detection and aggregate knowledge on different cases. Given that:
- Our users in general need to deal with YAML files a lot (so they are used to them)
- This particular file is not mandatory and is documented

I wouldn't consider this much of an issue. Maybe there's a better section to put documentation on it rather than the one I used?

A few considerations. The YAML file has no behavior, only static information to setup a unit test on a core function of the:
```console
$ spack external find ...
``` 
command. The logic for this unit test is the same for all packages. In some sense the information in the YAML file is of the same kind of what we can see in a parametrized fixture for other tests. The only difference is that here this information relates to different packages and is more verbose, so I decided to put it in a YAML file alongside the corresponding `package.py` rather than aggregating it in a single test file.

There's also a thing that worries me in taking the approach of an additional method. The package class is already overcharged with a lot of different responsibilities and I don't think it would help readability or maintenance to also add the responsibility of unit testing its own detection.

Reviewer: > I don't see this as a domain specific language, more like an additional YAML file that packagers can provide

> Our users in general need to deal with YAML files a lot (so they are used to them)

I think what you are asking is: “if we have config files, then why do you think this is bad?” Overall I would say the problem is that this doesn’t provide any of the benefits that config files typically provide to Spack users:

Users that write `detection_test.yaml` have to be familiar with the detection API and the specific implementation of the detection function in the associated `package.py` file. Users editing config files don’t need to know Spack internals, but users writing `detection_test.yaml` do.

Generally we keep config files as key/value stores or otherwise make them as simple as possible. Even the most complex config files we have in spack are entries which are collections of properties. `detection_test.yaml` specifies properties (like file locations) along with behavior (scripts) and outcomes (the expected spec output): that is more-involved than any config file we have. This also forces users to be familar with YAML syntax constraints which don’t appear in other config files, like the text block specification:

```
output: |
   echo "clang version 3.9.1-19ubuntu1 (tags/RELEASE_391/rc2)"
   echo "Target: x86_64-pc-linux-gnu"
   echo "Thread model: posix"
   echo "InstalledDir: /usr/bin"
```

Since the user has to understand (and in fact, coordinate with) the internals (i.e. the external detection function in `package.py`), I think this behavior should be specified as a function rather than a YAML file.

Author: > I think what you are asking is: “if we have config files, then why do you think this is bad?” Overall I would say the problem is that this doesn’t provide any of the benefits that config files typically provide to Spack users:

That's not the point. As I tried to explain above this YAML is not configuration - it's data fed into a unit test. I concur this part is true:

> Users that write detection_test.yaml have to be familiar with the detection API and the specific implementation of the detection function in the associated package.py file.

even though I don't see that as an issue, but then I don't understand the alternative solution of adding a method to the package. It shares the same charateristics (be familiar with the detection API and specific implementation of other functions in the package) AND it couples the package class with unit tests on it. Overall it seems to me no concern would be solved but more would be added.

**TL;DR** I pushed a modification in 55335df that moves the YAML files into a subfolder of the directory where we keep other unit test data files and moved the docs to the developer section. I hope this could be a good way to mitigate the concern you have over the user facing side of this PR and the one I have over adding an additional function to packages and coupling them to unit tests. The description at the top of the PR has been updated too.

Reviewer: > Overall it seems to me no concern would be solved but more would be added.

I want to avoid forcing users to switch between paradigms: if a user needs to be familiar with the external detection implementation of a package, then they will be reading Python. I don't want to force them to switch between Python and a YAML-based specification of behavior - I think they might as well write a Python function.

> I don't understand the alternative solution of adding a method to the package. It shares the same characteristics (be familiar with the detection API and specific implementation of other functions in the package)

> AND it couples the package class with unit tests on it

I think that is a good thing. They are already coupled, I don't see why separating them into different files changes that.

Put another way: what do you think is wrong with placing the testing logic in the package?

Author: To me the data in the YAML file is of the same kind as this:

https://github.com/spack/spack/blob/77a28b81ac10d63bca30c66edfd0205a11a1235e/lib/spack/spack/test/compilers/detection.py#L27-L38

so I'll counter a question with a question: do you think we should move that into the corresponding compiler class to be returned from a method:

```python
class Arm(Compiler):
    def input_for_version_detection(self):
        return [
            ('Arm C/C++/Fortran Compiler version 19.0 (build number 73) (based on LLVM 7.0.2)\n' # NOQA 
             'Target: aarch64--linux-gnu\n' 
             'Thread model: posix\n' 
             'InstalledDir:\n' 
             '/opt/arm/arm-hpc-compiler-19.0_Generic-AArch64_RHEL-7_aarch64-linux/bin\n', # NOQA 
             '19.0.0.73'), 
            ('Arm C/C++/Fortran Compiler version 19.3.1 (build number 75) (based on LLVM 7.0.2)\n' # NOQA 
             'Target: aarch64--linux-gnu\n' 
             'Thread model: posix\n' 
             'InstalledDir:\n' 
             '/opt/arm/arm-hpc-compiler-19.0_Generic-AArch64_RHEL-7_aarch64-linux/bin\n', # NOQA 
             '19.3.1.75') ]
```
?

Author: > what do you think is wrong with placing the testing logic in the package?

In the current form the unit test has knowledge of:
- The behavior of `spack external find` (or functions called within that module)
- The data needed to mock a filesystem structure and the expected result of a search

and drives the test - the package class doesn't know about the test, as it's not its responsibility.

With a method in the package class we have a unit test that calls a package that knows how the unit test is structured and returns the appropriate input - the unit test knows about the package and the package knows about the unit test and this makes the entire structure more coupled.

Being that 55335df makes it clear that the data is for unit tests and moves it away from packages, would that be good enough for you?

Reviewer: > so I'll counter a question with a question: do you think we should move that into the corresponding compiler class to be returned from a method:

Yes: all the compiler classes will eventually be packages anyways, so it would fit my expectation to move the example output into the compiler class (in the same way that I want to move the example package detection output into the package class).

> With a method in the package class we have a unit test that calls a package that knows how the unit test is structured and returns the appropriate input - the unit test knows about the package and the package knows about the unit test and this makes the entire structure more coupled.

I disagree with the first point of this and therefore with the conclusion, specifically you say:

> With a method in the package class we have a unit test that calls a package that knows how the unit test is structured and returns the appropriate input

Just because the package would need to follow a specific API to return example version output doesn't mean that it needs to know about the structure of the unit test. The API it needs to follow is:

```
class ExampleOutput(object):
  def __init__(self):
    self.expected_version = '1.0
    self.exe_to_output = {
      'bin/exe1': 'version 1.0',
      'bin/exe2': '1.0'
      ...
    }

class Package(object):
  def version_examples(self):
    # Return a set of ExampleOutput objects
   return []
```

But it wouldn't need to know how the unit test works

the ``paths`` attribute:

.. code-block:: yaml

   paths:
   - layout:
     - subdir: [bin]
       name: clang-3.9
       output: |
         echo "clang version 3.9.1-19ubuntu1 (tags/RELEASE_391/rc2)"
         echo "Target: x86_64-pc-linux-gnu"
         echo "Thread model: posix"
         echo "InstalledDir: /usr/bin"
     - subdir: [bin]
       name: clang++-3.9
       output: |
         echo "clang version 3.9.1-19ubuntu1 (tags/RELEASE_391/rc2)"
         echo "Target: x86_64-pc-linux-gnu"
         echo "Thread model: posix"
         echo "InstalledDir: /usr/bin"
     results:
     - spec: 'llvm@3.9.1 +clang~lld~lldb'

Each test is performed by first creating a temporary directory structure as
specified in the corresponding ``layout`` and by then running
package detection and checking that the outcome matches the expected
``results``. The exact details on how to specify both the ``layout`` and the
``results`` are reported in the table below:

.. list-table:: Test based on PATH inspections
   :header-rows: 1

   * - Option Name
     - Description
     - Allowed Values
     - Required
   * - ``layout``
     - Specifies the filesystem tree used for the test
     - List of objects

>      - Yes

Reviewer: What does "Yes" mean here?

Author: This part renders in the docs as a table. It describes in details the attributes that can be added to the YAML file:

![Screenshot from 2020-08-24 22-02-46](https://user-images.githubusercontent.com/4199709/91090581-9be05380-e655-11ea-9c85-78b9077c3924.png)

The "Yes" is under the "Required" column, to mean that the field is required. It's in fact a bit silly that now every field is required, but this was done thinking of future extensions to e.g. check that an external attribute is computed in a certain way. That wouldn't be required since not all packages have external attributes. If you think it's of no use at the moment I can remove the entire column.

> .. _package-lifecycle:

Reviewer 2: I think this may be in the wrong place now

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Add detection tests to packages
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> To ensure that a software is detected correctly for multiple configurations

Reviewer 2: ```suggestion
To ensure that software is detected correctly for multiple configurations
```

Author: Done in https://github.com/spack/spack/pull/18175/commits/d1c916b50af4517a2ee0b09f794fabceaedf3e2d

and on different systems users can write a ``detection_test.yaml`` file and

> put it in the same directory as the corresponding ``package.py`` file.

Reviewer 2: ```suggestion
put it in the packaage directory alongside the ``package.py`` file.
```

Author: Done in https://github.com/spack/spack/pull/18175/commits/d1c916b50af4517a2ee0b09f794fabceaedf3e2d

This YAML file contains enough information for Spack to mock an environment
and try to check if the detection logic yields the results that are expected.

As a general rule, attributes at the top-level of ``detection_test.yaml``

> represent search mechanisms and they all map to a list of tests that should confirm

Reviewer 2: ```suggestion
represent search mechanisms and they each map to a list of tests that should confirm
```

Author: Done in https://github.com/spack/spack/pull/18175/commits/d1c916b50af4517a2ee0b09f794fabceaedf3e2d

> the validity of each package detection logic.

Reviewer 2: ```suggestion
the validity of the package's detection logic.
```

Author: Done in https://github.com/spack/spack/pull/18175/commits/d1c916b50af4517a2ee0b09f794fabceaedf3e2d

""""""""""""""""""""""""""
Tests for PATH inspections

> """"""""""""""""""""""""""

Reviewer 2: Do we anticipate any other package detection method that we should be including tests for? We have the module-based detection for cray, but (a) I don't think we currently test it, because we can't test on the cray PE, and (b) I think we plan to deprecate that support as soon as we possibly can.

Author: Fwiw, we still have module based search for Cray and we might add other inspection method too - e.g. maybe querying system package managers etc.

Detection tests insisting on ``PATH`` inspections are listed under
the ``paths`` attribute:

.. code-block:: yaml

   paths:

>    - layout:
>      - subdir: [bin]
>        name: clang-3.9
>        output: |
>          echo "clang version 3.9.1-19ubuntu1 (tags/RELEASE_391/rc2)"
>          echo "Target: x86_64-pc-linux-gnu"
>          echo "Thread model: posix"
>          echo "InstalledDir: /usr/bin"
>      - subdir: [bin]
>        name: clang++-3.9
>        output: |
>          echo "clang version 3.9.1-19ubuntu1 (tags/RELEASE_391/rc2)"
>          echo "Target: x86_64-pc-linux-gnu"
>          echo "Thread model: posix"
>          echo "InstalledDir: /usr/bin"
>      results:
>      - spec: 'llvm@3.9.1 +clang~lld~lldb'

Reviewer 2: I'm worried that this format is very obtuse. Would we lose any information if we switched to something like:

```suggestion
   - layout:
     - executable: bin/clang-3.9
       output: |
        "clang version 3.9.1-19ubuntu1 (tags/RELEASE_391/rc2)"
        "Target: x86_64-pc-linux-gnu"
        "Thread model: posix"
        "InstalledDir: /usr/bin"
     - executable: bin/clang++-3.9
       output: |
        "clang version 3.9.1-19ubuntu1 (tags/RELEASE_391/rc2)"
        "Target: x86_64-pc-linux-gnu"
        "Thread model: posix"
        "InstalledDir: /usr/bin"
     specs:
     - 'llvm@3.9.1 +clang~lld~lldb'
```

This would seem to remove unnecessary verbosity.

Author: Done in https://github.com/spack/spack/pull/18175/commits/d40f73f6d71e9a1583411d306d957a264f50a05b
